cdc/log.cc

```diff
@@ -287,29 +304,35 @@ static schema_ptr create_log_schema(const schema& s, std::optional<utils::UUID>
             if (is_data_col) {
                 type = visit(*type, make_visitor(
                     // lists are represented as map<timeuuid, value_type>. Otherwise we cannot express delta
-                    [] (const list_type_impl& type) {
+                    [] (const list_type_impl& type) -> data_type {
                         return map_type_impl::get_instance(type.name_comparator(), type.value_comparator(), false);
                     },
-                    // user types are expressed as tuple<tuple<field0>, tuple<field1>...>
-                    // the extra tuple allows us to distinguish null == <no info>
-                    // from <null> = set field to null
-                    [] (const user_type_impl& type) -> data_type {
-                        std::vector<data_type> types;
-                        types.reserve(type.size());
-                        for (auto& ft : type.field_types()) {
-                            types.emplace_back(tuple_type_impl::get_instance({ ft }));
-                        }
-                        return tuple_type_impl::get_instance(std::move(types));
-                    },
                     // everything else is just frozen self
                     [] (const abstract_type& type) {
                         return type.freeze();
                     }
                 ));
-
-                type = tuple_type_impl::get_instance({ /* op */ data_type_for<column_op_native_type>(), /* value */ type->freeze()});
+                type = type->freeze();
             }
             b.with_column(log_data_column_name_bytes(column.name()), type);
+            b.with_column(log_data_column_deleted_name_bytes(column.name()), boolean_type);
+            if (column.type->is_multi_cell()) {
+                auto dtype = visit(*type, make_visitor(
+                    // lists are represented as map<timeuuid, value_type>. Otherwise we cannot express delta
+                    [] (const collection_type_impl& type) -> data_type {
+                        return set_type_impl::get_instance(type.name_comparator(), false);
+                    },
+                    // user types deletes are a set of the indices removed
+                    [] (const user_type_impl& type) -> data_type {
+                        return set_type_impl::get_instance(short_type, false);
+                    },
+                    // everything else is just frozen self
```

Comment:
This comment and the exception throw below somehow don't match in my mind. Is the comment really in the right place?

Reply:
doh. fixed.

Reply:
Thanks.